eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Deleg.hs

```diff
@@ -0,0 +1,214 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE EmptyDataDeriving #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module Cardano.Ledger.Conway.Rules.Deleg (
+  ConwayDELEG,
+  ConwayDelegEvent (..),
+  ConwayDelegPredFailure (..),
+) where
+
+import Cardano.Ledger.BaseTypes (
+  ShelleyBase,
+  StrictMaybe (SJust, SNothing),
+  maybeToStrictMaybe,
+ )
+import Cardano.Ledger.Binary (DecCBOR (..), EncCBOR (..))
+import Cardano.Ledger.Binary.Coders (
+  Decode (From, Invalid, SumD, Summands),
+  Encode (Sum, To),
+  decode,
+  encode,
+  (!>),
+  (<!),
+ )
+import Cardano.Ledger.Coin (Coin)
+import Cardano.Ledger.Conway.Core (ppKeyDepositL)
+import Cardano.Ledger.Conway.Era (ConwayDELEG)
+import Cardano.Ledger.Conway.TxCert (
+  ConwayDelegCert (ConwayDelegCert, ConwayRegCert, ConwayRegDelegCert, ConwayUnRegCert),
+  Delegatee (DelegStake, DelegStakeVote, DelegVote),
+ )
+import Cardano.Ledger.Core (Era (EraCrypto), EraPParams, EraRule)
+import Cardano.Ledger.Credential (Credential)
+import Cardano.Ledger.Keys (KeyRole (Staking))
+import Cardano.Ledger.Shelley.LedgerState (DState (..))
+import Cardano.Ledger.Shelley.Rules (DelegEnv (DelegEnv))
+import Cardano.Ledger.UMap (UView (RewDepUView))
+import qualified Cardano.Ledger.UMap as UM
+import Control.DeepSeq (NFData)
+import Control.Monad (unless)
+import Control.State.Transition (
+  BaseM,
+  Environment,
+  Event,
+  PredicateFailure,
+  STS,
+  Signal,
+  State,
+  TRC (TRC),
+  TransitionRule,
+  judgmentContext,
+  transitionRules,
+  (?!),
+ )
+import qualified Data.Map.Strict as Map
+import qualified Data.Set as Set
+import GHC.Generics (Generic)
+import Lens.Micro ((^.))
+import NoThunks.Class (NoThunks)
+
+data ConwayDelegPredFailure era
+  = IncorrectDepositDELEG
+      !(StrictMaybe Coin)
+  | StakeKeyAlreadyRegisteredDELEG
+      !(Credential 'Staking (EraCrypto era))
+  | StakeKeyNotRegisteredDELEG
+      !(Credential 'Staking (EraCrypto era))
+  | StakeKeyHasNonZeroAccountBalanceDELEG
+      !(Maybe Coin)
```

Comment:
Same here, having a predicate failure with non-zero balance implies that there is soma balance. It can't be Nothing, right?
```suggestion
  | StakeKeyHasNonZeroAccountBalanceDELEG
      !(Maybe Coin)
```